File: scripts/hex2psf1.py
#!/usr/bin/env python3

# Hex to PSF1 converter
# Because Unscii is offered in hex and I 
# already implemented PSF1
import re

glyphs = {}

with open('./other/unscii.hex') as f:
    for line in f:
        line = line.strip()
        if not line or ':' not in line:
            continue
        codepoint_str, hex_data = line.split(':', 1)
        codepoint = int(codepoint_str, 16)
        hex_data = re.sub(r'[^0-9a-fA-F]', '', hex_data)
        if codepoint < 256 and len(hex_data) >= 32:
            glyphs[codepoint] = bytes.fromhex(hex_data[:32])

with open('./other/unscii.psf', 'wb') as out:
    out.write(bytes([0x36, 0x04, 0x00, 16]))
    for i in range(256):
        out.write(glyphs.get(i, bytes([0] * 16)))
    size = out.tell()

print(f"Wrote {size} bytes")

File: scripts/hex2psf1.lua
#!/usr/bin/env lua
-- Converts hex fonts to psf1 fonts
-- checks for size, but still could
-- be errnonous. Whatever...

local glyphs = {}
local font_height = 16

for line in io.lines(arg[1]) do
   local code, hex = line:match("^(%x+)%s*:%s*(.+)$")
   if code then
      local bitmap = {}
      hex = hex:gsub("%s+", "")
      for i = 1, #hex, 2 do
         bitmap[#bitmap+1] = tonumber(hex:sub(i, i+1), 16) or 0
      end
      glyphs[tonumber(code, 16)] = bitmap
   end
end

local out = io.open(arg[2], "wb")
if not out then
   print("Cannot open " .. arg[2])
   os.exit(1)
end

out:write(string.char(0x36, 0x04, 0, font_height))

for i = 0, 255 do
   local g = glyphs[i]
   for j = 1, font_height do
      out:write(string.char(g and g[j] or 0))
   end
end

out:close()

local f = io.open(arg[2])
local size = f:seek("end")

f:close()

print(size == 4100 and "OK " .. size or "WRONG " .. size)
